src/backend/discovery/discovery.lambda.ts

```diff
@@ -0,0 +1,291 @@
+  sqs = new AWS.SQS();
+  functionContext = context;
+
+  if (!process.env?.STAGING_BUCKET_NAME) {
+    throw new Error('The STAGING_BUCKET environment variable is not set');
+  }
+  stagingBucket = stagingBucket;
+
+  if (!process.env?.QUEUE_URL) {
+    throw new Error('The NOTIFICATION_QUEUE_URL environment variable is not set');
+  }
+  queueUrl = process.env.QUEUE_URL;
+
+  AWS.config.update({
+    region: process.env.AWS_REGION,
+  });
+
+  // load marker from the staging bucket
+  let marker;
+  try {
+    const client = new AWS.S3();
+    await client.getObject({
+      Bucket: stagingBucket,
+      Key: MARKER_FILE_NAME,
+    }).promise().then(function(data) {
+      marker = data.Body?.toString();
+    });
+  } catch (error) {
+    console.log(`Failed to load marker for bucket: ${stagingBucket}, will start read from latest`);
+  }
+
+  console.log(`Starting changes stream read from ${marker}`);
+
+  const config: Nano.ChangesReaderOptions = {
+    includeDocs: true,
+    // pause the changes reader after each request
+    wait: true,
+    since: marker ?? 'now',
+    // `changesReader.get` stops once a response with zero changes is received, however it waits too long
+    //  since we want to terminate the Lambda function we define a timeout shorter than the default
+    timeout: TIMEOUT_MILLISECONDS,
+  };
+
+  const nano = Nano(NPM_REPLICA_REGISTRY_URL);
+  const db = nano.db.use('registry');
+
+  db.changesReader.get(config)
+    .on('batch', async (batch: Change[]) => {
+      console.log(`Received a batch of ${batch.length} length`);
+      const batchPromises = batch
+        // ignores changes which are not package update, see https://github.com/cdklabs/construct-hub/issues/3#issuecomment-858246275
+        .filter(change => change.doc.name)
+        .map(change => {
+          return processPackageUpdate(change)
+            .catch(err => {
+              Object.defineProperty(change, '_construct_hub_failure_reason', err);
+              return s3.putObject({
+                Bucket: stagingBucket,
+                Key: `${FAILED_CHANGE_PREFIX}/${change.id}`,
+                Body: JSON.stringify(change, null, 2),
+                Metadata: {
+                  'Lambda-Log-Group': functionContext.logGroupName,
+                  'Lambda-Log-Stream': functionContext.logStreamName,
+                  'Lambda-Run-Id': functionContext.awsRequestId,
+                },
+              }).promise().then(() => (void undefined));
+            });
+        });
+
+      await Promise.all(batchPromises).then(function() {
+        // write the last sequence to the marker file in S3
+        const lastSequence = batch[batch.length - 1].seq;
+        return writeMarkerToS3(lastSequence);
+      }).then(function() {
+        // resume reader
+        (db.changesReader as any).resume();
+      }).catch(function (error) {
+        throw Error(`Error while processing batch, marker will not be updated, exiting.\n ${error}`);
+      });
```

Comment:
Since we're doing `async...await`, there isn't a compelling reason to cascade callbacks using `.then()` this much... In this case, it takes less fluff to just write:

```suggestion
      try {
        await Promise.all(batchPromises);
      
        // write the last sequence to the marker file in S3
        const lastSequence = batch[batch.length - 1].seq;
        await writeMarkerToS3(lastSequence);

        // resume reader
        (db.changesReader as any).resume();
      } catch (error) {
        throw Error(`Error while processing batch, marker will not be updated, exiting.\n ${error}`);
      }
```